aprox network rewrite

maxpkatz opened the pull request.

zingale commented:
test suite: http://groot.astro.sunysb.edu/Microphysics/test-suite/gfortran/2022-02-19/index.html

Note: simplified-SDC doesn't work with this -- I think we need to ifdef out the linear algebra changes for now if we are not STRANG

zingale commented:
here are the Castro tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2022-02-19/index.html

the errors look reasonable, and most problems using these nets show nice speed ups (the one exception is wdmerger_collision, but that is built with `DEBUG=TRUE`)

zingale commented:
it looks like `interfaces/rhs_utilities.H` can be removed with this PR

zingale commented on networks/rhs.H:
```diff
@@ -0,0 +1,1604 @@
+    // Now count up all intermediate rates.
+    int count = 0;
+
+    for (int rate = 1; rate <= Rates::NumRates; ++rate) {
```
doesn't this double count an intermediate rate that is an additional rate to more than one rate?

maxpkatz replied:
This was definitely an issue I was concerned about, so my approach is to start with no reactions being intermediate (the counter is zero for all rates), and then we loop through all rates and increment the counter each a reaction is intermediate to some other reaction. It will be the case in general that some have a counter value > 1, but in the second loop where we actually count the number of intermediates, we're only checking for counter value > 0.

zingale commented on networks/rhs.H:
```diff
@@ -0,0 +1,1604 @@
+            fr   = fr * sc;
+        }
+
+        if constexpr (data.screen_reverse_reaction == 1) {
```
is this right?  For A + B -> C+ D, the reverse rate screening should use aion and zion from C and D

zingale commented on networks/rhs.H:
```diff
@@ -0,0 +1,1604 @@
+        frdt *= data.forward_branching_ratio;
+    }
+
+    if constexpr (data.forward_branching_ratio != 1.0_rt) {
```
should this be data.reverse_branching_ratio ?

maxpkatz replied:
Yes, fixed

zingale commented on networks/rhs.H:
```diff
@@ -0,0 +1,1604 @@
+
+    Real forward_term = fr;
+
+    if constexpr (data.species_A >= 0) {
```
can it really be == 0?  shouldn't this just be > 0?

maxpkatz replied:
I'm anticipating a future change where we do zero-indexing on the species, less code we'll have to change at that time.

zingale commented on networks/rhs.H:
```diff
@@ -0,0 +1,1604 @@
+
+            constexpr int exp = (spec2 == data.species_B) ? exponent - 1 : exponent;
+
+            static_assert(exp <= 3);
```
based on the assert in rhs_term, shouldn't this assert ensure that exp <= 2?

zingale replied:
oh, only if we are differenting wrt it.  I see the logic.  nevermind

maxpkatz commented:
> it looks like `interfaces/rhs_utilities.H` can be removed with this PR

Done

maxpkatz commented:
> Note: simplified-SDC doesn't work with this -- I think we need to ifdef out the linear algebra changes for now if we are not STRANG

I think this was just because I had forgotten to update numerical_jacobian.H to use the new interfaces when appropriate. This is now fixed.

zingale commented:
I'll retest
